### PR TITLE
Extract shopping logic to utils

### LINKED ISSUE #1841: Extract buying logic to utils
To make some more versatile and future components, extract all actions logic to `shoppingActions.ts`

https://github.com/kodadot/nft-gallery/blob/main-nuxt/components/rmrk/Gallery/AvailableActions.vue

Once it's done 
- #1840 could be made

### PR DESCRIPTION
**Pull request by Jarsen136**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).
👇 \_ Let's make a quick check before the contribution.

### PR type

- [ ] Bugfix
- [ ] Feature
- [x] Refactoring

### What's new?

- [x] PR closes #1841 
- [ ] <brief_description_of_what_I've_added>

---

 - [x] Refactor hardcoded actions into Interaction (from minimark)
  - [x] Refactor create remark to createInteraction (guess them import ;D)
  - [x] Refactor that mess in submit with MetaTransactionMixin (check other component how to use it)
  - [x] Use 🧹 and if you see some functions which are defined on the bad place, please put them in the correct file
  - [ ] Write unit tests for the functions and the AvailableAction.vue (file)

I could not complete the unit test for `AvailableAction.vue` because I'm having some weird issues with testing vue components : )

It would be great if someone could continue to write the unit test.

### Before submitting Pull Request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [ ] I've didn't break any original functionality
- [ ] I've posted a screenshot of demonstrated change in this PR

### Optional

- [x] I've tested it at </rmrk/collection/26902bc2f7c20c546a-1FVG7>
- [x] I've tested PR on mobile and everything seems works
- [x] I found edge cases
- [ ] I've written some unit tests 🧪

### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG)

### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

### Screenshot

- [ ] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -140,6 +141,7 @@
     "sass": "^1.51.0",
     "sass-loader": "^10.2",
     "typescript": "^4.6.4",
+    "vite-plugin-vue2": "^2.0.0",
```
Comment (roiLeo) [inline package.json:144]: already?

**Review by roiLeo — COMMENTED** [review]
Didn't test actions yet
- can't we use `AuthMixin` instead of passing `accountId` as a Prop()?
- we have another `AvailableActions` in `/components/unique/Gallery/item/`, wouldn't it be a good idea to change it too
- could you change `shoppingActions` relative path
- do we need vite plugin?
- not sure if `metaMixin` onError() callback will work

> I could not complete the unit test for AvailableAction.vue because I'm having some weird issues with testing vue components : )

What kind of issue?

File: package.json (hunk 1 above)
Comment (Jarsen136) [reply package.json:144]: if someone wants to write unit test for the vue component, he will need `@vue/test-utils` and `vite-plugin-vue2`

File: components/rmrk/Gallery/AvailableActions.vue [hunk 2]
```diff
@@ -27,25 +27,25 @@
       <template v-else-if="isForSale">
         <b-tooltip :active="buyDisabled" :label="$t('tooltip.buyDisabled')">
           <b-button
-            :type="iconType('BUY')[0]"
+            :type="iconType(`${ShoppingActions.BUY}`)[0]"
             :disabled="buyDisabled || !isAvailableToBuy"
             style="border-width: 2px"
             outlined
-            @click="handleAction('BUY')"
+            @click="handleAction(`${ShoppingActions.BUY}`)"
```
Comment (roiLeo) [inline components/rmrk/Gallery/AvailableActions.vue:34]:
```suggestion
            @click="handleAction(ShoppingActions.BUY)"
```

File: components/rmrk/Gallery/AvailableActions.vue [hunk 3]
```diff
@@ -27,25 +27,25 @@
       <template v-else-if="isForSale">
         <b-tooltip :active="buyDisabled" :label="$t('tooltip.buyDisabled')">
           <b-button
-            :type="iconType('BUY')[0]"
+            :type="iconType(`${ShoppingActions.BUY}`)[0]"
```
Comment (roiLeo) [inline components/rmrk/Gallery/AvailableActions.vue:30]:
```suggestion
            :type="iconType(ShoppingActions.BUY)[0]"
```

**Comment by roiLeo** [comment]
> > * we have another `AvailableActions` in `/components/unique/Gallery/item/`, wouldn't it be a good idea to change it too
> 
> I don't know why there are two almost the same components named 'AvailableActions'.

I think one is used for Rmrk and the other one for Statemine, Westmint, Bsx. it would be good to unify all this

For changes concerning vitest, I wouldn't do it in this pull request, feel free to open an issue with your problem

**Comment by vikiival** [comment]
> I think one is used for Rmrk and the other one for Statemine, Westmint, Bsx. it would be good to unify all this

Yup, agree :)

**Comment by Jarsen136** [comment]
> I think one is used for Rmrk and the other one for Statemine, Westmint, Bsx. it would be good to unify all this

Could we open another issue to handle this unification of these two components?
I think it's not in the description of the original issue and maybe will cost much more time : )

**Comment by vikiival** [comment]
> Could we open another issue to handle this unification of these two components?

File: package.json (hunk 1 above)
Comment (roiLeo) [reply package.json:144]: could you remove the 2 packages you added? Your tests pass without them
Capture d’écran 2022-05-06 à 4 58 25 PM

**Review by roiLeo — APPROVED** [review]
✅ lgtm

**Comment by prachi00** [comment]
@Jarsen136 resolve conflicts?

**Comment by Jarsen136** [comment]
> @Jarsen136 resolve conflicts?

Done. I have just solved the conflicts.

**Review by vikiival — CHANGES_REQUESTED** [review]

File: components/rmrk/Gallery/AvailableActions.vue [hunk 4]
```diff
@@ -106,22 +108,24 @@ const components = {
...
   @Prop() public price!: string
   @Prop() public nftId!: string
   @Prop({ default: () => [] }) public ipfsHashes!: string[]
   @Prop({ default: false }) public buyDisabled!: boolean
-  private selectedAction: Action = ''
+  private selectedAction: ShoppingActions | '' = ''
   private meta: string | number = ''
-  protected isLoading = false
-  protected status = ''
+  public isLoading = false
```
[the remark this answers is not in the document]
Comment (vikiival) [reply components/rmrk/Gallery/AvailableActions.vue:123]: So you can remove it :) then

[the remark this answers is not in the document]
Comment (Jarsen136) [reply components/rmrk/Gallery/AvailableActions.vue:123]: : ) oki

**Review by vikiival — APPROVED** [review]

**Comment by prachi00** [comment]
pay 200 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $200 @ 72.89 USD/KSM ~ 2.744 $KSM
🧗 [Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG](https://kusama.subscan.io/account/Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG)
🔗 [0x5768bd11a0e06bb79218b9522e3d9726a4a39d79059a30a3aa68aff28dab01de](https://kusama.subscan.io/extrinsic/0x5768bd11a0e06bb79218b9522e3d9726a4a39d79059a30a3aa68aff28dab01de)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Pull request merged**